### ISSUE TITLE
character_spacing bug after raising Prawn::Errors::CannotFit

### ISSUE THREAD

**Issue by ndbroadbent** [prawnpdf/prawn#1073]
I have some code that catches `Prawn::Errors::CannotFit` and just draws the text normally as a fallback. But there's a bug with `character_spacing`, where it is ignored after the `CannotFit` error,. However, it works if I change `character_spacing` to a new value. This also only happens when the error is raised inside a `pdf.transparent` block.

Here's a minimal POC to reproduce the issue:

```ruby
#!/usr/bin/env ruby
require 'prawn'

Prawn::Document.generate('/tmp/prawn-test.pdf') do |pdf|
  pdf.fill_color('000000')
  pdf.font 'Helvetica', size: 12, style: :normal

  pdf.text_box 'Lorem Ipsum Etcetera',
    at: [55, 633],
    width: 300,
    height: 15,
    overflow: :shrink_to_fit,
    character_spacing: 7

  pdf.font 'Courier', size: 18, style: :normal

  pdf.transparent(0.4) do
    begin
      pdf.text_box 'abc',
        at: [55, 570],
        width: 5,
        height: 5,
        overflow: :shrink_to_fit,
        character_spacing: 7
    rescue Prawn::Errors::CannotFit
      pdf.draw_text 'abc', at: [55, 570]
    end
  end

  pdf.font 'Helvetica', size: 14, style: :normal
  pdf.text_box 'Some More Lorem Ipsum Text',
    at: [55, 550],
    width: 350,
    height: 20,
    overflow: :shrink_to_fit,
    character_spacing: 7
end

`open /tmp/prawn-test.pdf`
```

Result:

<img width="414" alt="screen shot 2018-08-27 at 10 13 10 pm" src="https://user-images.githubusercontent.com/139536/44667970-6ee90e00-aa46-11e8-80bd-c8ae9e983016.png">

You can see that `character_spacing: 7` is not working for "Some More Lorem Ipsum Text".

------------------

If I remove the `pdf.transparent(0.4) do` wrapper by commenting out lines 17 and 28 (but still run the code inside the block), then I see the correct behavior:

<img width="564" alt="screen shot 2018-08-27 at 10 15 30 pm" src="https://user-images.githubusercontent.com/139536/44668111-b96a8a80-aa46-11e8-9f28-7e106444bf2f.png">

I can also see the correct behavior when I change line 38 to: `character_spacing: 8` (instead of 7). Changing the `character_spacing` to a different value might be clearing a cache somewhere.

Commenting out `pdf.draw_text 'abc', at: [55, 570]` in the rescue clause doesn't help.

Another "fix" is to move the `pdf.transparent(0.4) do` wrapper inside the `begin`...`rescue` block, like this:

```ruby
  begin
    pdf.transparent(0.4) do
      pdf.text_box 'abc',
        at: [55, 570],
        width: 5,
        height: 5,
        overflow: :shrink_to_fit,
        character_spacing: 7
    end
  rescue Prawn::Errors::CannotFit
    pdf.draw_text 'abc', at: [55, 570]
  end
```

But the `pdf.transparent` block never finishes, so the opacity gets stuck and affects all of the text that is added afterwards.

**Comment by ndbroadbent** [comment]
I've attempted to use `height_of` as a workaround. It calls `box.render(dry_run: true)`, so I assumed that it would be less "stateful" than calling `text_box`. Unfortunately that has the same issue.

```ruby
  pdf.transparent(0.4) do
    text_can_fit = true
    begin
      pdf.height_of('abc', at: [55, 570], width: 10, height: 10, overflow: :shrink_to_fit, character_spacing: 7)
    rescue Prawn::Errors::CannotFit
      text_can_fit = false
    end

    if text_can_fit
      pdf.text_box 'abc',
        at: [55, 570],
        width: 10,
        height: 10,
        overflow: :shrink_to_fit,
        character_spacing: 7
    else
      pdf.draw_text 'abc', at: [55, 570]
    end
  end
```

(This code reproduces the same issue.)

**Comment by ndbroadbent** [comment]
Ok, so I've figured out this only happens when the last `character_spacing` is exactly 7. I can change the other `character_spacing` values to any other number and am still able reproduce the issue. 

But if I change the "Some More Lorem Ipsum Text" `character_spacing` to any value that is not 7, then the issue stops happening.

I've also confirmed that the issue is present in Preview on Mac, Firefox, and Chrome (so it's probably not a PDF rendering issue.)
